### PR TITLE
Support sites deployed on paths other than "/" (by generating relative links)

### PR DESCRIPTION
**Pull request by eddgrant**
## What are you trying to do?

Make it possible to generate a fully working site, in which all intra-site links (i.e. links within the site, between pages, to assets etc) are generated as relative links, so that the site can be deployed at a path other than "/".

## Identifying a user need

At HMRC we use the tech-docs-template to deploy multiple docs-as-code sites at different paths on a single domain e.g.

```
docs.tax.service.gov.uk/site-1
docs.tax.service.gov.uk/site-2
docs.tax.service.gov.uk/site-3
... etc
```

Without these changes the site links do not function correctly.

## Is it a change or a bug fix?

Change.

## What’s changed?

The changes are driven via standard Middleman configuration:

```yaml
set :relative_links, true
activate :relative_assets
```

When enabled, the above configuration causes the following changes to happen:

1. The Table of Contents is generated using relative links (824116a9f3be15c9641e9047d638dca84ff5ed85)
2. The Table of Contents expands, collapsed and is highlighted using relative links (a2dde29337afcb84c7bd19ac800b0cf074072af7)
3. The search function works using relative links (a2dde29337afcb84c7bd19ac800b0cf074072af7)
4. The `/api/pages.json` endpoint returns relative links, such that the `tech-docs-monitor` can correctly generate page URLs when raising Slack notifications (16396e70a8db310ccbaca8121fbe9eb2bdb484bc). This requires a companion `tech-docs-monitor` change for which the PR is [here](https://github.com/alphagov/tech-docs-monitor/pull/37).

### Testing

We are testing these changes at HMRC at the moment and haven't yet observed anything which has broken as a result of the changes. We don't believe the changes cause any regressions when building sites in the normal way using absolute links. However we'd be grateful for any additional testing people can do.

### CODE REVIEW

**Comment by lfdebrux** [comment]
@eddgrant looks like there are some linter warnings, are you happy to fix these?

File: lib/assets/javascripts/_modules/search.js [hunk 1]
```diff
@@ -74,7 +76,8 @@
     }
 
     this.isOnSearchPage = function isOnSearchPage () {
-      return Boolean(window.location.pathname.match(/\/search(\/|\/index.html)?$/))
+      // .* prefix is used to match sites which use relative links.
+      return Boolean(window.location.pathname.match(/.*\/search(\/|\/index.html)?$/))
```
[the remark this answers is not in the document]
Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js:80]: That's a very good question! I thought it wasn't working without my amendment, however I wasn't aware that the match happened anywhere within the string. I will remove and test to see what happens.

[the remark this answers is not in the document]
Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js:80]: Have tested this and you're absolutely right, the ".*" is unnecessary. Will remove.

File: lib/assets/javascripts/_modules/search.js [hunk 2]
```diff
@@ -25,8 +26,9 @@
       $searchResults = $searchResultsWrapper.find('.search-results__content')
       $searchResultsTitle = $searchResultsWrapper.find('.search-results__title')
       $searchHelp = $('#search-help')
+      pathToSiteRoot = $element.data('pathToSiteRoot')
 
-      changeSearchAction()
+      changeSearchAction(pathToSiteRoot)
```
[the remark this answers is not in the document]
Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js]: Good spot, have removed the variable from the function call.

File: spec/table_of_contents/helpers_spec.rb [hunk 3]
```diff
@@ -175,7 +281,8 @@ def add_children(children)
       expect(subject.multi_page_table_of_contents(resources, current_page, config, current_page_html).strip).to eq(expected_multi_page_table_of_contents.strip)
     end
 
-    it "builds a table of contents from several page resources with a custom http prefix ocnfigured" do
+    #TODO: Assess whether this test is still relevant given that we are now generating relative links.
```
[the remark this answers is not in the document]
Comment (eddgrant) [reply spec/table_of_contents/helpers_spec.rb]: Great, have removed the TODO.

**Comment by eddgrant** [comment]
I've just realised the `spec` task doesn't run the Jasmine tests (which are failing). I'm not familiar with Jasmine but I'll see if I can figure out what's going on there.

**Comment by eddgrant** [comment]
Got it! I was using EC6 `const` variable declarations, which work fine in my browser, but I think Jasmine may be configured to build at a prior version to ES6 and didn't like the `const` declarations. Changing them to `var` declarations has done the trick, tests are now all passing.

File: lib/govuk_tech_docs/path_helpers.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      resource_path
+    end
+
+    def path_to_site_root(config, page_path)
+      if config[:relative_links]
+        number_of_ascents_to_site_root = page_path.to_s.split("/").reject(&:empty?)[0..-2].length
+        ascents = number_of_ascents_to_site_root.zero? ? ["."] : number_of_ascents_to_site_root.times.collect { ".." }
+        path_to_site_root = ascents.join("/").concat("/")
+      else
+        path_to_site_root = "/"
```
Comment (lfdebrux) [inline lib/govuk_tech_docs/path_helpers.rb:24]: Can we include `config[:http_prefix]` here, for cases where the user is using that instead of relative links?

File: lib/assets/javascripts/_modules/search.js [hunk 5]
```diff
@@ -65,7 +67,7 @@
       // We need JavaScript to do search, so if JS is not available the search
       // input sends the query string to Google. This JS function changes the
       // input to instead send it to the search page.
-      $searchForm.prop('action', '/search')
+      $searchForm.prop('action', pathToSiteRoot + 'search/index.html')
```
Comment (lfdebrux) [inline lib/assets/javascripts/_modules/search.js:70]: I think it's best if this doesn't have the `index.html` file name, to be consistent with other URLs. Is there a reason that you changed it?

```suggestion
      $searchForm.prop('action', pathToSiteRoot + 'search/')
```

Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js:70]: If we don't include the `index.html` then I _think_ we become reliant on the web server that the site is deployed behind being configured to have a default page value of `index.html`. Having said that I think there are other areas where we generate links which rely on the default page behaviour, so you could argue this is incongruous with that approach. Happy to try taking it out if you'd prefer consistency here.

File: lib/govuk_tech_docs/path_helpers.rb (hunk 4 above)
Comment (eddgrant) [reply lib/govuk_tech_docs/path_helpers.rb:24]: Could you give an example? I'm not quite sure what behaviour we'd be after here.

Comment (lfdebrux) [reply lib/govuk_tech_docs/path_helpers.rb:24]: I think, [based on what the `middleman-search` library does](https://github.com/manastech/middleman-search/blob/ed9e5254bb554988c755fc40238d9e9f67b20abe/lib/middleman-search/extension.rb#L29), that we just need to have `path_to_site_root` be `config[:http_prefix]`:

```suggestion
        path_to_site_root = config[:http_prefix]
```

I think `:http_prefix` should always be the path to the site root, even if the user hasn't defined it, and it should already default to `/`:

https://github.com/middleman/middleman/blob/97ddbc52eac7a937adcc7d418fa9c2fb4a0a1a1e/middleman-core/lib/middleman-core/application.rb#L154

Comment (eddgrant) [reply lib/govuk_tech_docs/path_helpers.rb]: Have just pushed a change to incorporate this.

File: lib/assets/javascripts/_modules/search.js [hunk 6]
```diff
@@ -149,7 +151,8 @@
         var content = s.processContent(result.content, query)
         output += '<li class="search-result">'
         output += '<h3 class="search-result__title">'
-        output += '<a href="' + result.url + '">'
+        var url = result.url.startsWith('/') ? result.url.slice(1) : result.url
+        output += '<a href="' + pathToSiteRoot + url + '">'
```
Comment (lfdebrux) [inline lib/assets/javascripts/_modules/search.js:155]: Ah, sorry, my previous suggestion broke things slightly 😅 

It turns out that the search index in `search.json` includes `http_prefix` in the URLs, so when using http prefix and not relative links we now end up with the http prefix duplicated 🤔 .

Comment (lfdebrux) [reply lib/assets/javascripts/_modules/search.js:155]: I think the easiest way to fix this is to check if `pathToSiteRoot` is a relative link, and only append if it is

```suggestion
        url = pathToSiteRoot.startsWith('..') ? pathToSiteRoot + url : '/' + url  // url already contains pathToSiteRoot when using :http_prefix
        output += '<a href="' + url + '">'
```

There might be a neater way to do that though 😅

Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js:155]: Aah, good spot @lfdebrux ! Will take a look. Out of interest, have you been able to get a site running locally when specifying `http_prefix`? All the sites I've tried break horribly as none of the assets are available under their prefixed URLs :thinking: .

Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js:155]: hmm, I'm a little confused by this. I _think_ it has been caused by this change in `path_helpers.rb`:

```ruby
path_to_site_root = config[:http_prefix]
```

After the change, this results in any configured `http_prefix` being added in the path to the site root, which is then fed down in to the search system, breaking the search.

[This (rather historical) comment](https://github.com/middleman/middleman/issues/484#issuecomment-6459631) seems to suggest that `http_prefix` is only intended for use to prefix assets (for example when using a CDN and needing to specify some prefix). Having read that I'm now feeling that our use of it to calculate the site root is contrary to its intended use. Unfortunately the issue that comment comes from was never resolved, despite nearly a year of discussion :disappointed:.

I think my preference would be to revert back to reporting the site root as `/` in `path_helpers.rb` when using absolute links, this feels more in keeping with the intent of `http_prefix` to me. Grateful for your thoughts @lfdebrux .

Comment (lfdebrux) [reply lib/assets/javascripts/_modules/search.js:155]: > Out of interest, have you been able to get a site running locally when specifying http_prefix? All the sites I've tried break horribly as none of the assets are available under their prefixed URLs 🤔 .

Yes, the assets do cause issues, however we recently closed one PR to fix that partially #197, and I have a PR open to fix that for govuk-frontend assets #292. There's also a fix needed for the path `search.json`, in #196, although I'm hoping that we can fix that for both http_prefix and relative_links in this PR 🤞. 

If you're interested in how I've configured everything, right now I have a couple of copies of the example site building together, with different configs, and I serve them from one directory using a Python test server.

In short what I'm doing is:

```
cd example
bundle exec middleman build --http-prefix=/with-http-prefix/
cd ../..
mkdir public
cp -r tech-docs-gem/example/build public/with-http-prefix
cd public
python3 -m http.server
```

That seems to work for me, if I have the right fixes in place. I think copying (or symbolic linking) the build into a folder with the same name as the http_prefix is important; it might also be important to make sure the http_prefix begins and ends with a slash, although I haven't played around with that too much.

Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js]: Sorry for the delay in replying @lfdebrux , thanks for the tips on getting things running with an `http_prefix`, that was really helpful. I have just pushed what I hope is an appropriate fix to ensure that `http_prefix` is no longer duplicated when we generate the links to the search result pages. I have tested it in all the use cases I can think of (relative links, absolute links with and without an http_prefix) but would be grateful if you could give it some :eyes: too. Cheers!

Comment (lfdebrux) [reply lib/assets/javascripts/_modules/search.js]: Awesome, thanks! I'll take a look 👀

**Comment by eddgrant** [comment]
I've just spotted the conflict arising between 3c629e753b480b85c4618ec850c32a42286c2e5a (recently merged to master) and 3e8770e3ad45ddeefceae688d6da5f84cae596e1 (my change). I'm not 100% sure yet but I think the change merged to master might be a more idiomatic as it seems to use a variable which is exposed by the `middleman-search` gem, rather than relying on my method of walking back to the site root.

I will do some testing to see if we can replace my change with the one already merged.

**Comment by eddgrant** [comment]
After inspection I can now see that the 2 commits do subtly different things. One provides the relative path to the site root and the other provides the path to the search index page. I have rebased my changes to include the other change and everything seems to be working.

If I push my rebased branch is it going to destroy all of our conversation above @lfdebrux ? :thinking: If it will (and that's a problem) then I could perhaps try pulling the changes in as a merge commit instead. Let me know what you'd rather?

**Comment by lfdebrux** [comment]
@eddgrant thanks for looking into it! A rebase and force push won't affect the conversation on this pull request page, so feel free to go ahead. It's preferred to a merge commit anyway.

File: lib/assets/javascripts/_modules/search.js [hunk 7]
```diff
@@ -51,7 +53,7 @@
     this.downloadSearchIndex = function downloadSearchIndex () {
       updateTitle('Loading search results')
       $.ajax({
-        url: searchIndexPath,
+        url: pathToSiteRoot + (searchIndexPath.startsWith('/') ? searchIndexPath.slice(1) : searchIndexPath),
```
Comment (lfdebrux) [inline lib/assets/javascripts/_modules/search.js:56]: Unfortunately this means we're getting `http_prefix/http_prefix` here as well.

I'd suggest we remove some of the complexity here and go back to just hardcoding `search.json`. Then we don't need both `data-search-index-path` and `data-path-to-site-root`.

```suggestion
        url: pathToSiteRoot + (searchIndexPath.startsWith('/') ? searchIndexPath.slice(1) : searchIndexPath),
```

File: lib/source/layouts/_search.erb [hunk 8]
```diff
@@ -1,5 +1,5 @@
 <% if config[:tech_docs][:enable_search] %>
-<div class="search" data-module="search" data-search-index-path="<%= search_index_path %>">
+<div class="search" data-module="search" data-path-to-site-root="<%= path_to_site_root config, current_page.path %>" data-search-index-path="<%= search_index_path %>">
```
Comment (lfdebrux) [inline lib/source/layouts/_search.erb:2]:
```suggestion
<div class="search" data-module="search" data-path-to-site-root="<%= path_to_site_root config, current_page.path %>">
```

**Review by lfdebrux — COMMENTED** [review]
I think this is almost ready to be merged <3, just one small suggested change 🙏

File: lib/assets/javascripts/_modules/search.js (hunk 7 above)
Comment (eddgrant) [reply lib/assets/javascripts/_modules/search.js:56]: Aah that's a shame. Happy to revert to a hardcoded `search.json`, however I should note we'll be undoing the changes made in https://github.com/alphagov/tech-docs-gem/commit/3c629e753b480b85c4618ec850c32a42286c2e5a. Is that ok?

Comment (lfdebrux) [reply lib/assets/javascripts/_modules/search.js:56]: Yep, fine by me.

**Comment by eddgrant** [comment]
@lfdebrux sorry for the delay, it's been a busy week. I've just pushed the change reverting back to always using `/search.json` for the search results.

**Comment by lfdebrux** [comment]
> @lfdebrux sorry for the delay, it's been a busy week. I've just pushed the change reverting back to always using `/search.json` for the search results.

No worries, thanks so much for all the time you've spent on this!

**Review by lfdebrux — APPROVED** [review]

**Comment by eddgrant** [comment]
No worries at all, likewise thanks for all your effort reviewing it! :pray:

**Comment by lfdebrux** [comment]
> Hey @lfdebrux I forgot to ask, do you have plans to tag this as a release and publish a new rubygem? I'm interested so I can plan when to deprecate our fork and move back to using this directly.

@eddgrant I've discussed this with the other working group members, we will try and put out a release soon.

**Comment by lfdebrux** [comment]
This issue has been fixed in [release v3.2.0](https://github.com/alphagov/tech-docs-gem/releases/tag/v3.2.0) 🎉 

You can now [configure your Tech Docs Template (TDT) to build your documentation site to use relative links to pages and assets](https://tdt-documentation.london.cloudapps.digital/configure_project/global_configuration/#build-your-site-using-relative-links).

Thanks [@eddgrant](https://github.com/eddgrant) for contributing this feature and the associated fixes.

This change was introduced in [pull request #291: Support sites deployed on paths other than "/" (by generating relative links)](https://github.com/alphagov/tech-docs-gem/pull/291).
